---
title: "scRNAseq Trout Immune Cells Pseudotime Analysis"
author: "Simone Oberhaensli, simone.oberhaensli@unibe.ch"
date: "2024-01-05"
output: html_document
---



```{r setup, echo=FALSE}
knitr::opts_chunk$set(
  comment = "#", echo = FALSE
)
# set your working directory
# setwd("05_james_pseudotime")
```


```{r libraries, include=FALSE}
# required libraries
# version 1.6.1
# library(Matrix, lib.loc = '.../05_james_pseudotime/Rpackages')

library(SeuratObject)
# visualization
library(ggplot2)
library(grDevices)
library(RColorBrewer)

# Monocle3
library(monocle3)
# library(SeuratWrappers)
library(patchwork)


# annoying issue with packages:
# SeuratObject/Seurat needs Matrix >= 1.6.3
# Monocle3/irlba needs Matrix 1.6.1
# as a solution I saved saved the monocle objects previously with Seurat and Matrix 1.6.4
# with Matrix v.1.6.5 it however didn't happen anymore?

# monocle needed an old version of Matrix, I installed it in a different directory 
# remotes::install_version("Matrix", version = "1.6-1.1")
# packageVersion("Matrix")
# custom installation of version 1.6.1
# > .libPaths()
# [1] "/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library"
# > myPaths <- .libPaths()
# > myPaths <- c(myPaths, '.../05_james_pseudotime/Rpackages')
# > .libPaths(myPaths)
# > .libPaths()
# [1] "/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library"    ".../05_james_pseudotime/Rpackages"
# then load the version
# remotes::install_version("Matrix", version = "1.6-1.1", lib='.../05_james_pseudotime/Rpackages')

```


### Summary

Run trajectory and pseudotime analysis on the scRNAseq trout immune cells data. 

Inspiration and resources:
[SIB course material 'Single-cell transcriptomics - trajectory analysis'](https://sib-swiss.github.io/single-cell-training/day3/day3-3_trajectory_analysis.html). 

Many thanks to the SIB for making these materials available to the community!

Monocole3 [Vignette](https://cole-trapnell-lab.github.io/monocle3/docs/introduction/). 


### Datasets

This analysis was done on an already pre-processed scRNAseq dataset. Available datasets were a) full dataset with 83k cells and b) downsampled dataset with 4k cells for testing. Both datasets are seurat objects.

```{r data_down, eval=FALSE}
# down sampled dataset
load("data/trout_SCTprocessed_PC30k30_var.ft.filt_161023.downsamp.rds")
dat <- immune.combined.sct.downsampled
rm(immune.combined.sct.downsampled)

cat("trout_SCTprocessed_PC30k30_var.ft.filt_161023.downsamp.rds loaded")

```


```{r data_full, eval=FALSE}
# the full dataset
load("data/trout_SCTprocessed_PC30k30_var.ft.filt_161023.rds")
dat <- immune.combined.sct
rm(immune.combined.sct)

cat("trout_SCTprocessed_PC30k30_var.ft.filt_161023.rds loaded")

```



### Monocle3

Monocle3 needs the seurat data to be converted into its own data format 'cell_data_set'. We do this for the already integrated data.

```{r load_data, eval=FALSE}
# create gene metadata data.frame
feature_names <- as.data.frame(rownames(dat))
rownames(feature_names) <- rownames(dat)
colnames(feature_names) <- "gene_short_name"


# original function: initiate monocle object from seurat count table 
# there are different options:
# assays: RNA, SCT and integrated
# immune.combined.sct@assays$integrated$scale.data
# immune.combined.sct@assays$integrated$data
seu_int_monocl <- monocle3::new_cell_data_set(Seurat::GetAssayData(dat, 
                                                                   assay = "integrated",
                                                                   layer = "data"),
                                              cell_metadata = dat@meta.data,
                                              gene_metadata = feature_names)

# run to save the object
# saveRDS(seu_int_monocl,file = "seu_int_monocl_fulldataset.RDS")

```

Preprocess the cells, here without normalization. Default is PCA, the elbow plot shows how much of the variance can be explained by the first five principal components


```{r clustering, eval=FALSE}
seu_int_monocl <- readRDS("seu_int_monocl_fulldataset.RDS")



seu_int_monocl <- monocle3::preprocess_cds(seu_int_monocl, num_dim = 50, norm_method = "none") 
# does work for the full dataset only if I don't normalize the data!
# Error with full dataset if we normalize:
# Warning: NaNs producedError in (function (A, nv = 5, nu = nv, maxit = 1000, work = nv + 7, reorth = TRUE,  : 
# max(nu, nv) must be positive


# elbow plot
monocle3::plot_pc_variance_explained(seu_int_monocl)

```


Run UMAP dimensionality reduction and color the cells according to the seurat UMAP clusters

```{r run_UMAP, eval=FALSE}
seu_int_monocl <- monocle3::reduce_dimension(seu_int_monocl, reduction_method = "UMAP")
# stdout: 
# No preprocess_method specified, using preprocess_method = 'PCA'
# Found more than one class "dist" in cache; using the first, from namespace 'spam'
# Also defined by ‘BiocGenerics’
# Found more than one class "dist" in cache; using the first, from namespace 'spam'
# Also defined by ‘BiocGenerics’




# plot the UMAP and colour like it was coloured before
p <- monocle3::plot_cells(seu_int_monocl, 
                     color_cells_by = "integrated_snn_res.0.3", 
                     cell_size = 0.5, 
                     show_trajectory_graph = FALSE)
p
# save UMAP
# ggsave("umap_colored_integrated_snn_res.0.3.pdf",p)


```

Cluster the cells

```{r cluster_cells, eval=FALSE}

seu_int_monocl <- monocle3::cluster_cells(seu_int_monocl, resolution = 0.00001) # resolution=0.00001 results in 24 clusters, 0.0005 in 39. Partitions stay the same
# stdout:
# No trajectory to plot. Has learn_graph() been called yet?

p1 <- monocle3::plot_cells(seu_int_monocl, label_cell_groups = F, cell_size = 0.5, show_trajectory_graph = FALSE)
#ggsave("clusters_00001.pdf",p1)

p2 <- monocle3::plot_cells(seu_int_monocl, color_cells_by = "partition",cell_size = 0.5, show_trajectory_graph = FALSE)
#ggsave("partitions_00001.pdf",p2)


```

Learn and plot the graph

```{r learn_graph, eval=FALSE}
seu_int_monocl <- monocle3::learn_graph(seu_int_monocl)
# save the learned graph
#saveRDS(seu_int_monocl, "seu_int_monocl_graph.rds")

seu_int_monocl <- readRDS("seu_int_monocl_graph.rds")
# stdout:
# Warning: no non-missing arguments to min; returning Inf

p4 <- monocle3::plot_cells(seu_int_monocl, cell_size = 0.5)
# ggsave("graph_allpartitions_00001.pdf",p4)

```

Now follows an interactive part.


```{r interactive_part, eval=FALSE}
# opens an interactive window and you are asked to choose your root node
seu_int_monocl <- monocle3::order_cells(seu_int_monocl)



p5 <- monocle3::plot_cells(seu_int_monocl,
           color_cells_by = "pseudotime",
           label_cell_groups=F,
           label_leaves=F,
           label_branch_points=FALSE,
           graph_label_size=1.5, cell_size = 0.5)
#ggsave("pseudotime_allclusters.pdf",p5)

# interactively choose a subset
seuB <- choose_cells(seu_int_monocl)
# seuB <- choose_graph_segments(seu_int_monocl) doesn't work...


p6 <- plot_cells(seuB, show_trajectory_graph = FALSE, cell_size = 0.5,
           color_cells_by = "pseudotime")
# stdout: 
# Message: Cells aren't colored in a way that allows them to be grouped.
# ggsave("pseudotime_bcells.pdf",p6)


```



```{r eof, eval=FALSE}

sessionInfo()

# Show in New Window
# Error in library(monocle3) : there is no package called ‘monocle3’
# Show in New Window
# R version 4.5.1 (2025-06-13)
# Platform: aarch64-apple-darwin20
# Running under: macOS Sequoia 15.6
# 
# Matrix products: default
# BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
# LAPACK: /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.1
# 
# locale:
# [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
# 
# time zone: Europe/Zurich
# tzcode source: internal
# 
# attached base packages:
# [1] stats4    stats     graphics  grDevices utils     datasets  methods   base     
# 
# other attached packages:
#  [1] shiny_1.10.0                monocle3_1.3.4              SingleCellExperiment_1.24.0 SummarizedExperiment_1.32.0 GenomicRanges_1.54.1        GenomeInfoDb_1.38.8         IRanges_2.36.0             
#  [8] S4Vectors_0.40.2            MatrixGenerics_1.14.0       matrixStats_1.5.0           Biobase_2.62.0              BiocGenerics_0.48.1         patchwork_1.3.0             ggplot2_3.5.2              
# 
# loaded via a namespace (and not attached):
#   [1] Rdpack_2.6.4            DBI_1.2.3               bitops_1.0-9            deldir_2.0-4            gridExtra_2.3           s2_1.1.9                rlang_1.1.6             magrittr_2.0.3         
#   [9] RcppAnnoy_0.0.22        e1071_1.7-16            compiler_4.5.1          reshape2_1.4.4          pbmcapply_1.5.1         systemfonts_1.2.3       vctrs_0.6.5             stringr_1.5.1          
#  [17] wk_0.9.4                pkgconfig_2.0.3         crayon_1.5.3            fastmap_1.2.0           XVector_0.42.0          labeling_0.4.3          promises_1.3.3          rmarkdown_2.30         
#  [25] nloptr_2.2.1            ragg_1.4.0              purrr_1.0.4             xfun_0.53               zlibbioc_1.48.2         cachem_1.1.0            jsonlite_2.0.0          later_1.4.2            
#  [33] DelayedArray_0.28.0     terra_1.8-54            parallel_4.5.1          biglm_0.9-3             R6_2.6.1                stringi_1.8.7           bslib_0.9.0             RColorBrewer_1.1-3     
#  [41] parallelly_1.45.0       boot_1.3-31             jquerylib_0.1.4         Rcpp_1.0.14             assertthat_0.2.1        knitr_1.50              httpuv_1.6.16           Matrix_1.7-3           
#  [49] splines_4.5.1           igraph_2.1.4            tidyselect_1.2.1        rstudioapi_0.17.1       dichromat_2.0-0.1       abind_1.4-8             yaml_2.3.10             viridis_0.6.5          
#  [57] codetools_0.2-20        listenv_0.9.1           lattice_0.22-7          tibble_3.3.0            plyr_1.8.9              withr_3.0.2             evaluate_1.0.5          future_1.40.0          
#  [65] sf_1.0-21               units_0.8-7             spData_2.3.4            proxy_0.4-27            pillar_1.10.2           KernSmooth_2.23-26      reformulas_0.4.1        generics_0.1.4         
#  [73] sp_2.2-0                RCurl_1.98-1.17         scales_1.4.0            minqa_1.2.8             globals_0.17.0          xtable_1.8-4            class_7.3-23            glue_1.8.0             
#  [81] slam_0.1-55             tools_4.5.1             speedglm_0.3-5          lme4_1.1-37             grid_4.5.1              spdep_1.3-13            tidyr_1.3.1             rbibutils_2.3          
#  [89] nlme_3.1-168            GenomeInfoDbData_1.2.11 cli_3.6.5               textshaping_1.0.1       S4Arrays_1.2.1          viridisLite_0.4.2       dplyr_1.1.4             gtable_0.3.6           
#  [97] sass_0.4.10             digest_0.6.37           classInt_0.4-11         SparseArray_1.2.4       ggrepel_0.9.6           farver_2.1.2            memoise_2.0.1           htmltools_0.5.8.1      
# [105] lifecycle_1.0.4         mime_0.13               MASS_7.3-65  


```


